### PR TITLE
Order Executor impl

### PR DESCRIPTION
**Pull request by oskin1**

### CODE REVIEW

File: spectrum-offchain/src/executor.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    /// Discard order in the case of fatal failure.
+    Fatal(String, TOrd::TOrderId),
+    /// Return order in the case of non-fatal failure.
+    NonFatal(String, TOrd),
+}
+
+pub trait RunOrder<TEntity, TCtx>: OnChainOrder + Sized {
+    /// Try to run the given `TOrd` against the given `TEntity`.
+    /// Returns transaction and the next state of the persistent entity in the case of success.
+    /// Returns `RunOrderError<TOrd>` otherwise.
```
Comment (kettlebell) [inline spectrum-offchain/src/executor.rs:28]: Name mismatch with `RunOrderFailure`. Also `RunOrderError` is a more idiomatic Rust name, I think

File: spectrum-offchain/src/executor.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                                })
+                                .await;
+                        }
+                    }
+                    Err(RunOrderFailure::NonFatal(err, ord)) => {
+                        warn!("Order suspended due to non-fatal error {}", err);
+                        self.backlog.suspend(ord).await;
+                    }
+                    Err(RunOrderFailure::Fatal(err, ord_id)) => {
+                        warn!("Order dropped due to non-fatal error {}", err);
```
Comment (kettlebell) [inline spectrum-offchain/src/executor.rs:90]: "...due to fatal error..."

**Review by kettlebell — APPROVED** [review]
Looks good to me, just some minor issues.

File: spectrum-offchain/src/executor.rs (hunk 1 above)
Comment (oskin1) [reply spectrum-offchain/src/executor.rs:28]: Fixed in the subsequent PR

File: spectrum-offchain/src/backlog.rs [hunk 3]
```diff
@@ -25,13 +25,15 @@ where
...
-    async fn suspend(&mut self, ord: SuspendedOrder<TOrd>) -> bool;
+    async fn suspend(&mut self, ord: TOrd) -> bool;
     /// Register successfull order to check if it settled later.
     async fn check_later(&mut self, ord: ProgressingOrder<TOrd>) -> bool;
     /// Pop best order.
     async fn try_pop(&mut self) -> Option<TOrd>;
     /// Remove order from backlog.
     async fn remove(&mut self, ord_id: TOrd::TOrderId);
+    /// Return order back to backlog.
+    async fn recharge(&mut self, org: TOrd);
```
[the remark this answers is not in the document]
Comment (oskin1) [reply spectrum-offchain/src/backlog.rs:36]: Fixed in the subsequent PR

File: spectrum-offchain/src/backlog.rs [hunk 4]
```diff
@@ -213,6 +223,19 @@ where
     async fn remove(&mut self, ord_id: TOrd::TOrderId) {
         self.store.drop(ord_id).await;
     }
+
+    async fn recharge(&mut self, org: TOrd) {
```
[the remark this answers is not in the document]
Comment (oskin1) [reply spectrum-offchain/src/backlog.rs:227]: Fixed in the subsequent PR

**Pull request merged**